File: merge-sort/java/MergeSort.java
public class MergeSort {
    
    public static void main(String[] args) {
        int[] data = {3, 2, 9, 5, 7, 1, 4, 6, 0, 8};
        System.out.print("before merge Sort: ");
        printArray(data);
        mergeSort(data);
        System.out.print("after merge Sort: ");
        printArray(data);
    }

    public static void mergeSort(int[] data) {
        mergeSortRecursion(data, 0, data.length - 1);
    }

    public static void mergeSortRecursion(int[] data, int left, int right) {
        if (left >= right)
            return;
        int mid = (left + right) / 2;
        mergeSortRecursion(data, left, mid);
        mergeSortRecursion(data, mid + 1, right);
        merge(data, left, mid, right);
    }

    public static void merge(int[] data, int left, int mid, int right) {
        int[] tempArray = new int[right - left + 1];
        int leftIndex = left;
        int rightIndex = mid + 1;
        int index = 0;
        while (leftIndex <= mid && rightIndex <= right) {
            if (data[leftIndex] <= data[rightIndex]) {
                tempArray[index] = data[leftIndex];
                leftIndex++;
            } else {
                tempArray[index] = data[rightIndex];
                rightIndex++;
            }
            index++; 
        }
        while (leftIndex <= mid) {
            tempArray[index] = data[leftIndex];
            leftIndex++;
            index++;
        }
        while (rightIndex <= right) {
            tempArray[index] = data[rightIndex];
            rightIndex++;
            index++;
        }
        for(int i=left; i<=right; i++) {
            data[i] = tempArray[i-left];
        }
    }

    public static void printArray(int[] data) {
        for(int i : data) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}

File: merge-sort/python/merge-sort.py
#!/usr/bin/python
#coding:utf8

def merge(list, left, mid, right):
    leftIndex = left
    rightIndex = mid + 1
    index = 0
    tempList = range(right - left + 1)
    while leftIndex <= mid and rightIndex <= right:
        if (list[leftIndex] <= list[rightIndex]):
            tempList[index] = list[leftIndex]
            leftIndex += 1
        else:
            tempList[index] = list[rightIndex]
            rightIndex += 1
        index += 1
    while leftIndex <= mid:
        tempList[index] = list[leftIndex]
        leftIndex += 1
        index += 1
    while rightIndex <= right:
        tempList[index] = list[rightIndex]
        rightIndex += 1
        index += 1
    for i in range(left, right + 1):
        list[i] = tempList[i - left]

def mergeSortRecursion(list, left, right):
    if left >= right:
        return;
    mid = (left + right) / 2
    mergeSortRecursion(list, left, mid)
    mergeSortRecursion(list, mid + 1, right)
    merge(list, left, mid, right)

def mergeSort(list):
    mergeSortRecursion(list, 0, len(list) - 1)

list = [4, 2, 8, 5, 7, 0, 3, 9, 1, 6]

print 'before merge sort: ', list

mergeSort(list)

print 'after merge sort: ', list


   
